### PR TITLE
add language model dataset

### PR DESCRIPTION
**Pull request by elephantmipt**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md), Pull Request section?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

## Description

I'm trying to make language modeling pipeline. I am planning to add minimal example also.
For now I think that something like:
```python
import torch
from transformers import AutoTokenizer, AutoModelWithLMHead
from catalyst.contrib.data.nlp import LMDataset
from catalyst.contrib.dl.runner import HuggingfaceRunner
from transformers.data.data_collator import DataCollatorForLanguageModeling

model = AutoModelWithLMHead.from_pretrained("distilbert-base-uncased")
tok = AutoTokenizer.from_pretrained("distilbert-base-uncased")

dataset = LMDataset(["this is a cool movie", "disgusting"], tok)
collate_fn = DataCollatorForLanguageModeling(tok).collate_batch
dataloader = torch.utils.data.DataLoader(dataset, collate_fn=collate_fn, batch_size=2)

optimizer = torch.optim.Adam(model.parameters())
loaders = {"train": dataloader}  # only train loader for LM

runner = HuggingfaceRunner()
    
runner.train(
    model=model,
    loaders=loaders,
    optimizer=optimizer,
    callbacks={
        "optimizer": dl.OptimizerCallback()
    },
    verbose=True
)
```
is a good idea.
## Related Issue

https://github.com/catalyst-team/catalyst/issues/805

## Type of Change

- [x] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: catalyst/contrib/data/nlp/language_modeling/dataset.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import warnings
+
+from tqdm.auto import tqdm
+
+import torch
+import transformers
+from transformers import AutoTokenizer
+
+
+class LMDataset(torch.utils.data.Dataset):
```
Comment (Scitator) [inline catalyst/contrib/data/nlp/language_modeling/dataset.py]: what do you think about full name usage, like `LanguageModelingDataset`?

File: catalyst/contrib/dl/runner/nlp/huggingface_runner.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from catalyst.dl import Runner
+from catalyst.dl.utils import is_wrapped_with_ddp
+
+
+class HuggingfaceRunner(Runner):
```
Comment (Scitator) [inline catalyst/contrib/dl/runner/nlp/huggingface_runner.py]: `HuggingFaceRunner` maybe?

File: catalyst/contrib/dl/runner/nlp/huggingface_runner.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    def _handle_batch(self, batch):
+        # output is always tuple in hf
+        # also if model is sepcified with task
+        # loss is stored in outpt[0]
+        loss = self.model(**batch)[0]
+
+        if is_wrapped_with_ddp(self.model):
+            loss = loss.mean()
+
+        self.state.batch_metrics["loss"] = loss
```
Comment (Scitator) [inline catalyst/contrib/dl/runner/nlp/huggingface_runner.py]: we also need model outputs into `state.output`

File: catalyst/contrib/dl/runner/nlp/huggingface_runner.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    """
+
+    def _handle_batch(self, batch):
+        # output is always tuple in hf
+        # also if model is sepcified with task
+        # loss is stored in outpt[0]
+        loss = self.model(**batch)[0]
+
+        if is_wrapped_with_ddp(self.model):
+            loss = loss.mean()
```
Comment (Scitator) [inline catalyst/contrib/dl/runner/nlp/huggingface_runner.py]: not sure if we need it, Catalyst already have `distributed_mean` for metrics

File: requirements/requirements-nlp.txt [hunk 5]
```diff
@@ -1 +1 @@
-transformers<=2.4.0>=2.0.0
\ No newline at end of file
+transformers>=2.8.0
```
Comment (Scitator) [inline requirements/requirements-nlp.txt]: why could not we use older versions of `transformers` ?

Comment (Scitator) [reply requirements/requirements-nlp.txt]: `2.8.0` is... very new :)
why could not we work with `2.0.0` still?

**Review by Scitator — CHANGES_REQUESTED** [review]
btw, I think it's better to split this PR into 2:
1. for Dataset - easy contribution
2. for Runner and examples  - need to discuss

File: catalyst/contrib/dl/runner/nlp/huggingface_runner.py (hunk 4 above)
Comment (PUSSYMIPT) [reply catalyst/contrib/dl/runner/nlp/huggingface_runner.py:17]: I get loss with shape (2,) if model is DataParallel, so I can't do backward. Is this a bug?

File: requirements/requirements-nlp.txt (hunk 5 above)
Comment (elephantmipt) [reply requirements/requirements-nlp.txt]: I need `DataCollatorForLanguageModeling` from transformers. It was published about 20 days ago. Maybe I can just add this class to utils with credentials to hf and use older version.

File: catalyst/contrib/dl/runner/nlp/huggingface_runner.py (hunk 4 above)
Comment (elephantmipt) [reply catalyst/contrib/dl/runner/nlp/huggingface_runner.py]: I get output shape of (2,) if i use DataParallel. So I need reduction because I obviously can't do backward. Is it a bug?

Comment (Scitator) [reply catalyst/contrib/dl/runner/nlp/huggingface_runner.py]: looks like so (not sure about huggingface internals)

File: requirements/requirements-nlp.txt (hunk 5 above)
Comment (Scitator) [reply requirements/requirements-nlp.txt]: I suppose, we could add `version` check
something like https://github.com/catalyst-team/catalyst/blob/master/catalyst/utils/seed.py#L25

**Comment by Scitator** [comment]
nevertheless, I still suggest to split your PR into several

File: requirements/requirements-nlp.txt (hunk 5 above)
Comment (elephantmipt) [reply requirements/requirements-nlp.txt]: well, there is literally nothing like this before 2.8.0, so I can't do something like
```python
if parse(transformers.__version__) >= Version("2.8.0"):
    collate_fn = DataCollatorForLanguageModeling().collate_batch
else:
    # what should I do?)
    collate_fn = utils.DataCollatorForLanguageModeling().collate_batch   # ?
```
also I need `max_len` arg in PreTrainedTokenizer. I'm not sure about version, but this feature is pretty new too.

File: catalyst/contrib/data/nlp/language_modeling/dataset.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            ]
+            if sort:
+                self.encoded.sort(key=len)
+
+        elif sort:
+            warnings.warn(
+                "Warning! lazy set to True so we can't sort"
+                + " sequences by length.\n"
+                + "You should set sort=False if lazy=True"
+            )
```
Comment (Scitator) [inline catalyst/contrib/data/nlp/language_modeling/dataset.py]: maybe, you mean `raise Exception`?

File: catalyst/contrib/data/nlp/language_modeling/dataset.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        If lazy is True then encode text and return
+        else just return already encoded text
+        """
+        if not self.lazy:
+            return torch.tensor(self.encoded[idx])
+
+        encoded = self.tokenizer.encode(
+            self.texts[idx], max_len=self.max_seq_length
+        )
+        return torch.tensor(encoded)
```
Comment (Scitator) [inline catalyst/contrib/data/nlp/language_modeling/dataset.py]: could we redefine `__getitem__` behaviour into `__init__`?

for example:
```
class:
  init_fn:
    self.__getitem__ = self.getitem_lazy if lazy else self.getitem_encode

  getitem_lazy_fn:
    return ...

  getitem_encode:
    return ...
```

File: catalyst/contrib/data/nlp/language_modeling/tests/test_dataset.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+
+def test_tokenizer_str():
+    """Test initialization with string"""
+    dataset = LanguageModelingDataset(texts, "bert-base-uncased")
+    assert dataset[0] is not None
+    assert len(dataset) == 2
+
+
+def test_tokenizer_tokenizer():
```
Comment (Scitator) [inline catalyst/contrib/data/nlp/language_modeling/tests/test_dataset.py:21]: btw, could we also add test for `exception raise`?

**Review by Scitator — CHANGES_REQUESTED** [review]

**Comment by Scitator** [comment]
btw, don't you mind to also update the CHANGELOG?

File: catalyst/contrib/data/nlp/language_modeling/dataset.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                self.encoded.sort(key=len)
+
+        elif sort:
+            raise Exception(
+                "lazy is set to True so we can't sort"
+                " sequences by length.\n"
+                "You should set sort=False if lazy=True"
+                " if you want to encode text in __get_item__ function"
+            )
+        self.length = len(texts)
```
Comment (Scitator) [inline catalyst/contrib/data/nlp/language_modeling/dataset.py]: btw, could we raise exception, or maybe assert in the beginning of initialisation?
currently, we could get to the situation, when we have encoded some dataset (big one) and then caught and exception - that's not cool :)

File: catalyst/contrib/data/nlp/language_modeling/tests/test_dataset.py (hunk 8 above)
Comment (elephantmipt) [reply catalyst/contrib/data/nlp/language_modeling/tests/test_dataset.py:21]: done
